weather.1
.\" Manpage for location
.\" Contact wilfried.meindl@gmail.com to correct errors or typos.
.\" Sections: NAME, SYNOPSIS, DESCRIPTION, OPTIONS, RETURN VALUES
.\"           ENVIRONMENT, FILES, EXAMPLES, DIAGNOSTICS, ERRORS,
.\"           SEE ALSO, STANDARDS, BUGS, SECURITY CONSIDERATIONS,
.\"           other
.TH man 1 "12 November 2015" "1.0" "weather man page"
.SH NAME
weather \- display the current weather conditions and a five day forecast
.SH SYNOPSIS
.SY weather
.OP \-h
.OP \-c
.OP \-f rcfile
.OP \-u unit
.I [woeid|place]
.SH DESCRIPTION
.B location
queries the Yahoo! database for information about a geographical place.
\fIlocation\fP is a string specifying what to search for.  It can
contain a city or country name, a zip code etc.  Also combinations like
"Springfield, Missoury" are possible.  It displays the name of the
location, the state, the country, the latitude, the longitude and the
WOEID (Where On Earth IDentifier).  A request can have multiple results.
.SH OPTIONS
.TP
.B \-h
Print the results in a more human readable format.
.TP
.BI "\-s " separator
Specifies the field separator.
It has no effect if the \fB\-h\fP option is used.
.SH ENVIRONMENT
.TP
.B AWKLIB
The path where the awk files are found.
By default set to $HOME/.awklib.
See also section
.BR FILES .
.SH FILES
.TP
getxml.awk
The XML parsing library.
.TP
location.awk
The parsing routines for location data.
.SH EXAMPLES
1. Search for New York.
Display the result in human readable format.
.PP
.EX
.RS 4
$ location -h "New York"
1. New York, New York, United States
        Zip: 10007
        Latitude: 40.71455, Longitude: -74.00714
        WOEID: 2459115
.RE
.EE
.PP
2. Display the same result in easily parsable form.
.PP
.EX
.RS 4
$ location "New York"
1|New York|New York|United States|10007|40.71455|...
.RE
.EE
.PP
3. Now with tabs as delimiters.
.PP
.EX
.RS 4
$ location -s "\\t" "New York"
1	New York	New York	United States	10007...
.RE
.EE
.PP
4. A query with multiple results
.PP
.EX
.RS 4
$ location Springfield
1|Springfield|Missouri|United States|65806|37.20897|...
2|Springfield|Massachusetts|United States|01103|...
3|Springfield|Illinois|United States|62701|39.80106|...
4|Springfield|Ohio|United States|45502|39.92503|...
\&...
.RE
.EE
.SH SEE ALSO
weather(1)
.SH BUGS
No known bugs.
.SH AUTHOR
Wilfried Meindl <wilfried.meindl@gmail.com>


weather.awk
#!/usr/bin/awk -f /home/wilfried/.awklib/getxml.awk -f

BEGIN {
  while (getXML(ARGV[1], 1)) {
    if (XTYPE == "TAG" && XITEM ~ /^yweather:/) {
      tagName = XITEM;
      sub(/^yweather:/, "", tagName);
      if (tagName == "forecast") {
	forecast += 1;
	for (attributeName in XATTR)
	  data[tagName,forecast,attributeName] = XATTR[attributeName];
      } else
	  for (attributeName in XATTR)
	    data[tagName,attributeName] = XATTR[attributeName];
    }
    if (XERROR) {
      print "Parsing error"
      exit 1
    }
  }

  if (computer_readable) {
      OFS=";"
      print "location", data["location", "city"], data["location", "region"],
            data["location", "country"]
      print "units", data["units", "temperature"], data["units", "distance"],
            data["units", "pressure"], data["units", "speed"]
      print "condition", data["condition", "date"], data["condition", "temp"],
            data["condition", "text"], data["condition", "code"]
      print "wind", data["wind", "chill"], data["wind", "direction"], 
            wind_direction(data["wind", "direction"]), data["wind", "speed"]
      print "atmosphere", data["atmosphere", "humidity"],
            data["atmosphere", "visibility"], data["atmosphere", "pressure"],
            data["atmosphere", "rising"]
      print "astronomy", data["astronomy", "sunrise"],
            data["astronomy", "sunset"]
      for (i = 1; i <= 5; i++) {
          print "forecast_" i, data["forecast", i, "day"],
                data["forecast", i, "date"], data["forecast", i,"low"],
                data["forecast", i, "high"], data["forecast", i, "text"],
                data["forecast", i, "code"]
      }
  } else {
      printf "Weather report for %s, %s, %s:\n", data["location", "city"],
             data["location", "country"], data["condition", "date"]
      printf "Condition:\t%s °%s, %s\n", data["condition", "temp"],
             data["units", "temperature"], data["condition", "text"]
      printf "Wind:\t\t%s %s %s, %s °%s\n", data["wind", "speed"],
             data["units", "speed"], wind_direction(data["wind", "direction"]),
             data["wind", "chill"], data["units", "temperature"]
      printf "Atmosphere:\t%s%%, %s %s %s\n", data["atmosphere", "humidity"],
             data["atmosphere", "pressure"], data["units", "pressure"],
             pressure_direction(data["atmosphere", "rising"])
      printf "Sun:\t\t%s - %s\n", data["astronomy", "sunrise"],
             data["astronomy", "sunset"]
      printf("Forecast:\n")
      for (i = 1; i <= 5; i++) {
          printf("\t%s, %s: %s °%s - %s °%s, %s\n",
                 data["forecast", i, "day"], data["forecast_" i, "date"],
                 data["forecast", i, "low"], data["units", "temperature"],
                 data["forecast", i, "high"], data["units", "temperature"],
                 data["forecast", i, "text"])
      }
  }
}

function pressure_direction(n) {
  if (n == 0) return "steady"
  else
    if (n == 1) return "falling"
  else
    if (n == 2) return "rising"
}

function wind_direction(n) {
  if      ((  0    <= n) && (n <  11.25)) return "N"
  else if (( 11.25 <= n) && (n <  33.75)) return "NNE"
  else if (( 33.75 <= n) && (n <  56.25)) return "NE"
  else if (( 56.25 <= n) && (n <  78.75)) return "ENE"
  else if (( 78.75 <= n) && (n < 101.25)) return "E"
  else if ((101.25 <= n) && (n < 123.75)) return "ESE"
  else if ((123.75 <= n) && (n < 146.25)) return "SE"
  else if ((146.25 <= n) && (n < 168.75)) return "SSE"
  else if ((168.75 <= n) && (n < 191.25)) return "S"
  else if ((191.25 <= n) && (n < 213.75)) return "SSW"
  else if ((213.75 <= n) && (n < 236.25)) return "SW"
  else if ((236.25 <= n) && (n < 258.75)) return "WSW"
  else if ((258.75 <= n) && (n < 281.25)) return "W"
  else if ((281.25 <= n) && (n < 303.75)) return "WNW"
  else if ((303.75 <= n) && (n < 326.25)) return "NW"
  else if ((326.25 <= n) && (n < 348.75)) return "NNW"
  else if ((348.75 <= n) && (n <= 360))   return "N"
}
